src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Security/Permissions/DefaultPermissionService.cs

```diff
@@ -0,0 +1,58 @@
+namespace OrchardCore.Security.Permissions;
+
+public sealed class DefaultPermissionService : IPermissionService
+{
+    private readonly IEnumerable<IPermissionProvider> _permissionProviders;
+
+    // Cached per request.
+    private Dictionary<string, Permission> _permissions;
+
+    public DefaultPermissionService(IEnumerable<IPermissionProvider> permissionProviders)
+    {
+        _permissionProviders = permissionProviders;
+    }
+
+    public async ValueTask<Permission> FindByNameAsync(string name)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(name);
+
+        if (_permissions is null)
+        {
+            await LoadPermissionsAsync();
+        }
+
+        if (_permissions.TryGetValue(name, out var permission))
+        {
+            return permission;
+        }
+
+        return null;
+    }
+
+    public async ValueTask<IEnumerable<Permission>> GetPermissionsAsync()
```

Comment:
If you return `IReadOnlyList<Permission>` all the `ToArray()` can be removed. And the result can be arrays or lists. Same for ByNames.

Reply:
The .ToArray() are needed convert IEnumerable<Permission> to IList<string> I don't think there is a benefit on using IReadOnlyList unless I am missed something. I did some cleanup now to use the `FindByNamesAsync` extension in more places.